### PR TITLE
Create `metabase.permissions.util` namespace

### PR DESCRIPTION
**Pull request by noahmoss**
I was getting annoyed at how bloated `metabase.models.permissions` was becoming. It was basically a catch-all namespace for everything permissions-related, which made it hard to find functions or know where to make changes.

In this PR I've created a top-level `permissions` directory where we can put perms-related namespaces, and I've started it off with `metabase.permissions.util`, with a chunk of code (largely) copied over from `metabase.models.permissions`. Also created the corresponding test namespace and copied some tests over.

No functionality should be changed here, it's just a step towards getting things more organized.

### CODE REVIEW

**Comment by noahmoss** [comment]
Current dependencies on/for this PR:
* `master`
  * **PR #37177** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37177?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
    * **PR #37178** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37178?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/metabase/metabase/37177?utm_source=stack-comment).

File: src/metabase/models/permissions.clj [hunk 1]
```diff
@@ -929,13 +721,15 @@
                                                            [:like :object (h2x/literal "%/db/%")]])
                                 (m/map-vals (fn [paths]
                                               ;; remove v1 paths, implicitly keep v2 paths
-                                              (remove (fn [path] (mc/validate [:re (u.regex/rx "^/" v1-data-permissions-rx "$")]
+                                              (remove (fn [path] (mc/validate [:re (u.regex/rx
+                                                                                    "^/"
+                                                                                    perms.u/v1-data-permissions-rx
+                                                                                    "$")]
```
Comment (escherize) [inline src/metabase/models/permissions.clj:727]: you didn't change this, but I noticed it's rebuilding 

```
(u.regex/rx "^/" v1-data-permissions-rx "$")
```

on each iteration of remove

File: test/metabase/api/card_test.clj [hunk 2]
```diff
@@ -2943,10 +2944,10 @@
                                            uploads-database-id (mt/id)
                                            uploads-table-prefix nil
                                            uploads-schema-name "PUBLIC"]          (let [{:keys [status body]} (upload-example-csv-via-api!)]
-            (is (= 200
-                   status))
-            (is (= body
-                   (t2/select-one-pk :model/Card :database_id (mt/id)))))))
+                                                                                   (is (= 200
```
Comment (escherize) [inline test/metabase/api/card_test.clj:2947]: this let belongs on the newline, I think.

File: test/metabase/permissions/util_test.clj [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+                          w w     w w       w w     w w
+                          w w           w           w w
+                          w w           w           w w
+                          w w w w w w w w w w w w w w w;
+                              w w w w w w w w w w w;;;;
+                              w w w w w w w w w w w;;
+                                  w w w w w w w;
+                                      w   w;
+                                      w   w
+                                    w w   w w))))));
```
Comment (escherize) [inline test/metabase/permissions/util_test.clj:302]: 🦉

**Review by escherize — APPROVED** [review]

File: test/metabase/api/card_test.clj (hunk 2 above)
Comment (noahmoss) [reply test/metabase/api/card_test.clj:2947]: oops, how'd that happen

File: src/metabase/models/permissions.clj (hunk 1 above)
Comment (noahmoss) [reply src/metabase/models/permissions.clj:727]: Changing to use `perms.u/DataPath`

**Pull request merged**